controller_manager/test/test_controller_manager.cpp

```diff
@@ -104,8 +104,11 @@ TEST_P(TestControllerManagerWithStrictness, controller_lifecycle)
     lifecycle_msgs::msg::State::PRIMARY_STATE_UNCONFIGURED, test_controller->get_state().id());
 
   // configure controller
-  cm_->configure_controller(test_controller::TEST_CONTROLLER_NAME);
-  cm_->configure_controller(TEST_CONTROLLER2_NAME);
+  {
+    ControllerManagerRunner cm_runner(this);
```

Comment:
why did you add those runners here?

Reply:
Yeah! This is something I didn't like, but this is needed. Once we load the controllers and run one update cycle to check that the unconfigured controllers are not being updated as shown in lines L97-L101. 

The CM update method uses the method `rt_controllers_wrapper_.update_and_get_used_by_rt_list()` to get the updated list, then the internal variable `used_by_realtime_controllers_index_` of the `RTControllerListWrapper` gets updated to the proper index. So, as a result of this, once I sort all the configured controllers at the time of switching using the method CM L792 `rt_controllers_wrapper_.switch_updated_list(guard);` it gets blocked in the method  `wait_until_rt_not_using`, as the `used_by_realtime_controllers_index_` still points to the same value. Until the `update_and_get_used_by_rt_list` is called, it is not updated, and this usually happens in the update cycle. That's the main reason to use these runners.

https://github.com/ros-controls/ros2_control/blob/e0291041ec2ebd3020139c3bb6d3ce6036c3e02c/controller_manager/src/controller_manager.cpp#L2137-L2148

https://github.com/ros-controls/ros2_control/blob/e0291041ec2ebd3020139c3bb6d3ce6036c3e02c/controller_manager/src/controller_manager.cpp#L2155-L2166

https://github.com/ros-controls/ros2_control/blob/e0291041ec2ebd3020139c3bb6d3ce6036c3e02c/controller_manager/src/controller_manager.cpp#L2104-L2108

https://github.com/ros-controls/ros2_control/blob/e0291041ec2ebd3020139c3bb6d3ce6036c3e02c/controller_manager/src/controller_manager.cpp#L2030-L2034

If I remove the first call of the update cycle to check the unconfigured controllers doesn't increase the internal counter, then I wouldn't need these Runners. However, I think it is wise to leave those checks for not to make mistakes in the future. If not, what I can do is, I can remove this initial update cycle check in most and leave it in one or two places, that way we can remove some runners. What do you think @destogl ?

Reply:
This is good! Thanks for the explanation – I forgot about update of the counters.